devfile/devworkspace-operator
# Migrate to devfile/api v1alpha2 and add conversion webhooks.

### What does this PR do?
Switches the devworkspace-operator to use `v1alpha2` devworkspaces, and implements conversion webhooks. Currently this is done by updating `go.mod` to point at https://github.com/devfile/api/pull/213, as that PR is not yet merged.

To avoid too many temporary changes in the repo (changing where devworkspace CRDs are downloaded from), you'll have to manually copy in the CRDs generated in https://github.com/devfile/api/pull/213 to `config/crd/bases/`: 
```bash
curl "https://raw.githubusercontent.com/devfile/api/fc53138e/crds/workspace.devfile.io_devworkspaces.yaml" \
    -o config/crd/bases/workspace.devfile.io_devworkspaces.yaml

curl "https://raw.githubusercontent.com/devfile/api/fc53138e/crds/workspace.devfile.io_devworkspacetemplates.yaml" \
    -o config/crd/bases/workspace.devfile.io_devworkspacetemplates.yaml 
```

### What issues does this PR fix or reference?
Resolves https://github.com/devfile/devworkspace-operator/issues/189

### Is it tested? How?
To test this PR (on minikube, for now):
0. Set env vars as usual for development
1. Deploy cert-manager on your cluster: `make install_cert_manager`
2. Deploy controller: `make docker install`
3. Verify that conversion webhooks are created: log message `conversion webhook enabled` appears in `kubectl logs -f deploy/devworkspace-webhook-server`
4. Create a DevWorkspace: `kubectl apply -f samples/theia.yaml`
5. Check that you can get both v1alpha1 and v1alpha2 versions of the devworkspace:
    ```
    kubectl get devworkspaces.v1alpha1.workspace.devfile.io theia -o yaml
    kubectl get devworkspaces.v1alpha2.workspace.devfile.io theia -o yaml
    ```
6. Test usability as normal

### Remaining TODOs
1. I haven't tested/verified that local run/debug still works as expected, but will do that soon.
2. These changes make webhooks mandatory; we likely need to update how `WEBHOOKS_ENABLED` works, perhaps disabling the mutating and validating webhooks when set to false but still deploying the webhooks server
3. The service name for the webhooks server is hard-coded, so it has to be hard-coded into templates as well, we should probably make this configurable
4. Only k8s with cert-manager is supported at this point. The next step is to add support for OpenShift via the [Service CA operator](https://github.com/openshift/service-ca-operator), which supports a similar CRD annotation. We'll need to verify that the certificates secret is correctly propagated to the webhooks server
5. This PR includes a workaround for deploying CRDs when they're already present on the cluster (we first remove them). I've found that on `minikube`, without those changes, the second `make install` fails. You can test this by
    ```bash
    kustomize build config/crd | kubectl apply -f -
    kustomize build config/crd | kubectl apply -f -
    ```
    or by commenting out the delete command in `make install`.
6. This PR updates *existing* CRDs for Components and WorkspaceRoutings. In terms of yaml spec they are identical, except the new version has defaults for some endpoint fields. We need to verify that 
    - Applying a `v1` version of the CRD is not an error when a `v1beta1 CRD with the same GVK is present
    - That upgrading from a previous version will not cause issues for existing devworkspaces.
